File: Python/Easy/Filter/Array/2176__Count_Equal_and_Divisible_Pairs_in_an_Array.py
# 1 - method

# def countPairs(nums: list[int], k: int) -> int:
#     n = len(nums)
#     count = 0
#     for i in range(n-1):
#         for j in range(i+1, n):
#             if nums[i] == nums[j] and (i * j) % k == 0:
#                 count += 1
#     return count
#
#
# nums = [3, 1, 2, 2, 2, 1, 3]
# k = 2
# print(countPairs(nums, k))

from math import gcd
from collections import defaultdict


def countPairs(nums: list[int], k: int) -> int:
    num_to_indices = defaultdict(list)
    count = 0

    for i, num in enumerate(nums):
        for j in num_to_indices[num]:
            if (i * j) % k == 0:
                count += 1
        num_to_indices[num].append(i)
    return count


nums = [3, 1, 2, 2, 2, 1, 3]
k = 2
print(countPairs(nums, k))

File: PHP/Easy/Filters/Array/2176__Count_Equal_and_Divisible_Pairs_in_an_Array.php
<?php
function countPairs($nums, $k)
{
    $len = count($nums);
    $count = 0;
    for ($i = 0; $i < $len - 1; $i++) {
        for ($j = $i + 1; $j < $len; $j++) {
            if ($nums[$i] == $nums[$j] and ($i * $j) % $k == 0) {
                $count++;
            }
        }
    }
    return $count;
}
$nums = [3, 1, 2, 2, 2, 1, 3];
$k = 2;